File: ios/Classes/CASMethodHandler.swift
//
//  CASMethodHandler.swift
//  Runner
//
//  Created by Dmytro Uzhva on 12.08.2024.
//

import CleverAdsSolutions
import Flutter

private let logTag = "CASMethodHandler"

class CASMethodHandler: MethodHandler {
    func onMethodCall(call: FlutterMethodCall, result: @escaping FlutterResult) -> Bool {
        switch call.method {
        case "getSDKVersion": getSDKVersion(result: result)
        case "validateIntegration": validateIntegration(result: result)
        default: return false
        }
        return true
    }

    private func getSDKVersion(result: @escaping FlutterResult) {
        result(CAS.getSDKVersion())
    }

    private func validateIntegration(result: @escaping FlutterResult) {
        CAS.validateIntegration()
        result(nil)
    }
}

File: android/src/main/kotlin/com/cleveradssolutions/plugin/flutter/CASMethodHandler.kt
package com.cleveradssolutions.plugin.flutter

import android.app.Activity
import com.cleversolutions.ads.android.CAS
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.Result

private const val LOG_TAG = "CASMethodHandler"
private const val CHANNEL_NAME = "com.cleveradssolutions.plugin.flutter/cas"

class CASMethodHandler(
    private val activityProvider: () -> Activity?
) : MethodHandler(CHANNEL_NAME) {

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "getSDKVersion" -> getSDKVersion(result)
            "validateIntegration" -> validateIntegration(result, activityProvider)
            else -> super.onMethodCall(call, result)
        }
    }

    private fun getSDKVersion(result: Result) {
        result.success(CAS.getSDKVersion())
    }

    private fun validateIntegration(result: Result, activityProvider: () -> Activity?) {
        activityProvider()?.let {
            CAS.validateIntegration(it)
            result.success(null)
        } ?: result.error(LOG_TAG, "Method: 'validateIntegration', Activity is null", null)
    }

}